packager/packager.cc

```diff
@@ -725,46 +795,122 @@ class ImportTree final {
-                addImport(importedPackage, import.name.loc, exp.fqn, import.type);
+    void mergeImports(core::Context ctx, const PackageInfoImpl &package) {
+        const auto &packageDB = ctx.state.packageDB();
+        UnorderedMap<core::NameRef, core::LocOffsets> importedNames;
+        for (auto &import : package.importedPackageNames) {
+            auto &imported = import.name;
+            auto &importedPackage = packageDB.getPackageInfo(imported.mangledName);
+            if (!importedPackage.exists()) {
+                if (auto e = ctx.beginError(imported.loc, core::errors::Packager::PackageNotFound)) {
+                    e.setHeader("Cannot find package `{}`", imported.toString(ctx));
+                }
+                continue;
+            }
+
+            if (importedNames.contains(imported.mangledName)) {
+                if (auto e = ctx.beginError(imported.loc, core::errors::Packager::InvalidImportOrExport)) {
+                    e.setHeader("Duplicate package import `{}`", imported.toString(ctx));
+                    e.addErrorLine(core::Loc(ctx.file, importedNames[imported.mangledName]),
+                                   "Previous package import found here");
+                }
+            } else {
+                importedNames[imported.mangledName] = imported.loc;
+
+                // TODO (aadi-stripe): fix this so it's not quadratic (if it ends up causing runtime bloat). Currently
+                // (11/8/2021) timing analysis suggests that it doesn't.
+                //
+                // Determine whether the current package either imports a suffix of the imported name, or itself is a
+                // suffix of the imported name. This bit is passed into mergeImportsFromImportedPackage, specifically
+                // into the "enumerateExportsInFull" parameter.
+                //
+                // To save on memory usage by reducing the number of nodes created in the AST, we individually alias
+                // exported constants from the imported package only when there would be an import conflict otherwise -
+                // i.e., when this is either a subpackage of the imported name or imports a subpackage of the imported
+                // name.
+                const bool isOrImportsSubpackage =
+                    absl::c_any_of(package.importedPackageNames,
+                                   [&](const auto &otherImport) -> bool {
+                                       return otherImport.name.fullName.isSuffix(imported.fullName);
+                                   }) ||
+                    package.name.fullName.isSuffix(imported.fullName);
+                mergeImportsFromImportedPackage(ctx, PackageInfoImpl::from(importedPackage), import,
+                                                isOrImportsSubpackage);
             }
         }
     }
 
     // Make add imports for the test package for all normal code exported by its corresponding
     // "normal" package.
```

Comment:
(Okay, this one _does_ already have a comment, but I am having a hell of a time actually trying to parse this sentence. …not that you need to fix it in this PR: I'm just commenting.)

Reply:
I can clean this up — not the most clear comment.